sam/assembly.go

```diff
@@ -0,0 +1,219 @@
+	"math"
+)
+
+type DiploidBase byte
+
+const (
+	AA DiploidBase = 0
+	AC DiploidBase = 1
+	AG DiploidBase = 2
+	AT DiploidBase = 3
+	CC DiploidBase = 4
+	CG DiploidBase = 5
+	CT DiploidBase = 6
+	GG DiploidBase = 7
+	GT DiploidBase = 8
+	TT DiploidBase = 9
+	NN DiploidBase = 10
+)
+
+// diploidBaseString formats a DiploidBase struct as a string for debugging.
+func diploidBaseString(base DiploidBase) string {
+	switch base {
+	case AA:
+		return "AA"
+	case AC:
+		return "AC"
+	case AG:
+		return "AG"
+	case AT:
+		return "AT"
+	case CC:
+		return "CC"
+	case CG:
+		return "CG"
+	case CT:
+		return "CT"
+	case GG:
+		return "GG"
+	case GT:
+		return "GT"
+	case TT:
+		return "TT"
+	case NN:
+		return "NN"
+	}
+	log.Fatalf("Unrecognized DiploidBase: %v.", base)
+	return ""
+}
+
+//DiploidBaseCallFromPile determines which of 10 DiploidBase genotypes
+// are present at the position of an input Pile, preconditioned on the assumption of
+// a diploid base at that position (in other words, not considering INDELs).
+// epsilon represents the error rate, a parameter for the likelihood function.
+func DiploidBaseCallFromPile(p Pile, refBase dna.Base, priorCache [][]float64, homozygousCache [][]float64, heterozygousCache [][]float64, epsilon float64) DiploidBase {
+	if refBase == dna.N { //we will force a return of diploid NN whenever an N is found in the reference.
+		return NN
+	}
+	var aCount = p.CountF[dna.A] + p.CountR[dna.A]
+	var cCount = p.CountF[dna.C] + p.CountR[dna.C]
+	var gCount = p.CountF[dna.G] + p.CountR[dna.G]
+	var tCount = p.CountF[dna.T] + p.CountR[dna.T]
+	var baseCount = aCount + cCount + gCount + tCount
+
+	if baseCount < 1 {
+		switch refBase {
+		case dna.A:
+			return AA
+		case dna.C:
+			return CC
+		case dna.G:
+			return GG
+		case dna.T:
+			return TT
+		case dna.N: //should be redundant with catch above
+			return NN
+		default:
+			log.Fatalf("Reference base was not N, A, C, G, or T. Found: %s.\n", dna.BaseToString(refBase))
+		}
+	}
+
+	var maxDiploid []DiploidBase
+	var maxPosterior float64
+
+	switch refBase {
+	case dna.A:
+		maxDiploid = []DiploidBase{AA}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, AA, epsilon, homozygousCache, heterozygousCache), priorCache[dna.A][AA])
+	case dna.C:
+		maxDiploid = []DiploidBase{CC}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, CC, epsilon, homozygousCache, heterozygousCache), priorCache[dna.C][CC])
+	case dna.G:
+		maxDiploid = []DiploidBase{GG}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, GG, epsilon, homozygousCache, heterozygousCache), priorCache[dna.G][GG])
+	case dna.T:
+		maxDiploid = []DiploidBase{TT}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, TT, epsilon, homozygousCache, heterozygousCache), priorCache[dna.T][TT])
+	default:
+		log.Fatalf("Reference base was not N, A, C, G, or T. Found: %s.\n", dna.BaseToString(refBase))
```

Comment:
You may consider a panic here (and elsewhere in this file) instead of a fatal. This is a pretty handy function I can see being used in other places in gonomics. If this is used outside of samAssembly tracking down this error could be pretty hard. A stack trace from a panic would help a lot.

Reply:
This is a great idea. Will implement in next commit